### PR TITLE
.Net: Handlebars Planner 

### PR DESCRIPTION
**Pull request by teresaqhoang**
### Motivation and Context

This pull request adds a new sequential planner based on the Handlebars Templating language.  Having a simpler scripting language for the LLM to produce yields better results and you can do more complex things (like loops, complex object support, and more). 

### Description

The changes include adding a Handlebars planner, a HandlebarsPlan class, and a HandlebarsTemplateEngineExtensions class to support the use of Handlebars templates in generating plans. 

Additionally, the pull request adds Handlebars template engine extensions and system helpers, as well as several new Handlebars helpers and functions for semantic planning. The changes also include updates to the console application documentation, improvements to error handling, and the addition of a new HandlebarsPlanner KernelSyntaxExample.

Changes:
- Added Handlebars planner, HandlebarsPlan class, and HandlebarsTemplateEngineExtensions class to support Handlebars templates in generating plans
- Added Handlebars template engine extensions and system helpers for more flexible and customizable planning and rendering of prompts and responses
- Added several built-in Handlebars helpers and functions for semantic planning, including greaterThan, lessThanOrEqual, greaterThanOrEqual, json, message, raw, doubleOpen, doubleClose, set, and get
- Implemented error handling in HandlebarsPlanner for cases where available functions cannot create a plan for the given goal
- Added several new HandlebarsPlanner examples and tests to demonstrate usage

Examples:
image

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

image

### CODE REVIEW

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlan.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+using System.Text.Json.Serialization;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace Microsoft.SemanticKernel.Planners.Handlebars;
+
+/// <summary>
+/// Represents a Handlebars plan.
+/// </summary>
+public sealed class HandlebarsPlan
```
Comment (lemillermicrosoft) [inline dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlan.cs:13]: Have you considered having this implement `ISKFunction`?

**Comment by joe-chiu** [comment]
Found via https://github.com/microsoft/semantic-kernel/discussions/3429

I wonder if a more python like syntax would be easier for LLM to generate as many coding fine tuned LLM has more innate training / understanding for such syntax and would require less handholding via prompt?

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+/// <summary>
+/// Represents a Handlebars planner.
+/// </summary>
+public class HandlebarsPlanner : IHandlebarsPlanner
+{
+    /// <summary>
+    /// The key for the available kernel functions.
+    /// </summary>
+    public static readonly string AvailableKernelFunctionsKey = "AVAILABLE_KERNEL_FUNCTIONS";
```
Comment (TaoChenOSU) [inline dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs:24]: nit: use "const" instead of "static readonly"

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    /// </summary>
+    /// <param name="kernel">The kernel.</param>
+    /// <param name="config">The configuration.</param>
+    public HandlebarsPlanner(IKernel kernel, HandlebarsPlannerConfig? config = default)
+    {
+        this._kernel = kernel;
+        this._config = config ?? new HandlebarsPlannerConfig();
+    }
+
+    private readonly HashSet<ParameterType> _parameterTypes = new();
```
Comment (TaoChenOSU) [inline dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs:44]: nit: put this at the top with the other class members

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            {
+                case "user~":
+                    chatMessages.AddUserMessage(message);
+                    break;
+                case "system~":
+                    chatMessages.AddSystemMessage(message);
+                    break;
+                case "assistant~":
+                    chatMessages.AddAssistantMessage(message);
+                    break;
```
Comment (TaoChenOSU) [inline dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs:118]: default case?

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+
+/// <summary>
+/// Provides extension methods for rendering Handlebars templates in the context of a Semantic Kernel.
+/// </summary>
+public static class HandlebarsTemplateEngineExtensions
+{
+    /// <summary>
+    /// The key used to store the reserved output type in the dictionary of variables passed to the Handlebars template engine.
+    /// </summary>
+    public static readonly string ReservedOutputTypeKey = "RESERVED_OUTPUT_TYPE";
```
Comment (TaoChenOSU) [inline dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs:25]: nit: const

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,405 @@
...
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <returns>The rendered Handlebars template.</returns>
+    public static string Render
+    (
+        IKernel kernel,
+        SKContext executionContext,
+        string template,
+        Dictionary<string, object?> variables,
+        CancellationToken cancellationToken = default
+    )
```
Comment (TaoChenOSU) [inline dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs:43]: nit: positions of the opening and closing parenthesizes.
```suggestion
public static string Render(
        IKernel kernel,
        SKContext executionContext,
        string template,
        Dictionary<string, object?> variables,
        CancellationToken cancellationToken = default)
```

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+                if (arguments[0].GetType() == typeof(HashParameterDictionary))
+                {
+                    // Process hash arguments
+                    var handlebarArgs = arguments[0] as IDictionary<string, object>;
+
+                    // Prepare the input parameters for the function
+                    foreach (var param in functionView.Parameters)
+                    {
+                        // TODO: accomodate ServerUrl override? 
+                        if (param.Name.Contains("server_url") || param.Name.Contains("server-url"))
```
Comment (TaoChenOSU) [inline dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs:88]: This can be removed after this PR: https://github.com/microsoft/semantic-kernel/pull/3427

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs (hunk 4 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs:123]: A default case is not needed here because we don't want to parse out any other kind of messages except these certain ones. Others would be considered a failed case and cause the template to throw an exception when rendered.

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs (hunk 7 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs]: Awesome! Removed!

File: dotnet/samples/KernelSyntaxExamples/Example65_HandlebarsPlanner.cs [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        }
+
+        [SKFunction, SKName("GetDefinition"), System.ComponentModel.Description("Gets the definition for a given word.")]
+        public string GetDefinition([System.ComponentModel.Description("Word to get definition for.")] string word)
+        {
+            return this._dictionary.TryGetValue(word, out var definition)
+                ? definition
+                : "Word not found";
+        }
+    }
```
Comment (alliscode) [inline dotnet/samples/KernelSyntaxExamples/Example65_HandlebarsPlanner.cs:215]: nit: Can we have an example that shows a plugin that uses complex types? It's OK if this comes in a follow up PR.

File: dotnet/samples/KernelSyntaxExamples/Program.cs [hunk 9]
```diff
@@ -21,7 +21,7 @@ public static async Task Main(string[] args)
         using CancellationTokenSource cancellationTokenSource = new();
         CancellationToken cancelToken = cancellationTokenSource.ConsoleCancellationToken();
 
-        string? defaultFilter = null; // Modify to filter examples
+        string? defaultFilter = "Example65"; // Modify to filter examples
```
Comment (alliscode) [inline dotnet/samples/KernelSyntaxExamples/Program.cs:24]: Should this change be checked in?

File: dotnet/src/Planners/Planners.Core/Handlebars/ParameterType.cs [hunk 10]
```diff
@@ -0,0 +1,104 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Json.Serialization;
+
+namespace Microsoft.SemanticKernel.Planners.Handlebars;
+
+internal class Property
```
Comment (alliscode) [inline dotnet/src/Planners/Planners.Core/Handlebars/ParameterType.cs:10]: 'Property` is pretty generic, could this be called `HandlebarsProperty`? Similar for ParameterType.

**Review by alliscode — APPROVED** [review]

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+using Microsoft.SemanticKernel.Diagnostics;
+using Microsoft.SemanticKernel.Orchestration;
+using Microsoft.SemanticKernel.Text;
+
+namespace Microsoft.SemanticKernel.Planners.Handlebars;
+
+/// <summary>
+/// Provides extension methods for rendering Handlebars templates in the context of a Semantic Kernel.
+/// </summary>
+public static class HandlebarsTemplateEngineExtensions
```
Comment (teresaqhoang) [inline dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs:20]: @markwallace-microsoft I was hoping to integrate this with the TemplateEnginer.Handlebars you created but we need more support for registering the built in helpers. I'll create an ADR for this.

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlan.cs [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+
+        // try
+        // {
+        //     variables.TryGetValue(HandlebarsPromptTemplateEngineExtensions.ReservedOutputTypeKey, out var returnType);
+        //     var deserializedResult = JsonSerializer.Deserialize(results, (returnType as Type) ?? typeof(string));
+        //     results = deserializedResult != null ? deserializedResult.ToString() : results;
+        // }
+        // catch (Exception)
+        // {
+        // }
```
Comment (TaoChenOSU) [inline dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlan.cs:64]: Is this comment block left here intentionally?

Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlan.cs]: Yeah I left it there because it's part of the complex types integration but I'll remove it for this PR

File: dotnet/src/Planners/Planners.Core/Handlebars/ParameterType.cs (hunk 10 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/ParameterType.cs]: I changed it to "NestedProperty" for now but I'm going to remove it with the work to integrate your type support changes altogether

File: dotnet/samples/KernelSyntaxExamples/Example65_HandlebarsPlanner.cs (hunk 8 above)
Comment (teresaqhoang) [reply dotnet/samples/KernelSyntaxExamples/Example65_HandlebarsPlanner.cs:215]: Yep will come in follow-up PR

File: dotnet/src/Planners/Planners.Core/Handlebars/skPrompt.handlebars [hunk 13]
```diff
@@ -0,0 +1,138 @@
+{{#message role="system"}}## Instructions
```
Comment (markwallace-microsoft) [inline dotnet/src/Planners/Planners.Core/Handlebars/skPrompt.handlebars:1]: Should the filename be `skprompt.txt`?

File: dotnet/src/Planners/Planners.Core/Handlebars/IHandlebarsPlanner.cs [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.SemanticKernel.Diagnostics;
+
+namespace Microsoft.SemanticKernel.Planners.Handlebars;
+
+/// <summary>
+/// Interface for planner that uses a set of semantic functions to create a sequential plan.
+/// </summary>
+public interface IHandlebarsPlanner
```
Comment (markwallace-microsoft) [inline dotnet/src/Planners/Planners.Core/Handlebars/IHandlebarsPlanner.cs:13]: Are we expecting multiple different implementations?

Feels like should have an `IPlanner` interface which uses generics. I think we should look at this before we make planners v1.0.

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs (hunk 11 above)
Comment (markwallace-microsoft) [reply dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs:18]: Can this be made internal?

File: dotnet/src/Planners/Planners.Core/Handlebars/IHandlebarsPlanner.cs [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+{
+    /// <summary>
+    /// Gets the stopwatch used for measuring planning time.
+    /// </summary>
+    Stopwatch Stopwatch { get; }
```
Comment (markwallace-microsoft) [inline dotnet/src/Planners/Planners.Core/Handlebars/IHandlebarsPlanner.cs:18]: Should this be part of the interface?

**Review by markwallace-microsoft — COMMENTED** [review]
Can we reduce the public API surface to just `HandlebarsPlanner` and `HandlebarsPlan`?

We should also remove any public interfaces we don't expect customers to implement e.g. `IHandlebarsPlanner`.

File: dotnet/src/Planners/Planners.Core/Handlebars/skPrompt.handlebars (hunk 13 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/skPrompt.handlebars:1]: No, I intentionally made the extension .handlebars to have the formatting benefits in the IDE
image

Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/skPrompt.handlebars:1]: I believe this is all changing with the new semantic function template format @matthewbolanos is proposing, but in the meantime, I can change it to skprompt.txt for consistency if you think that's best?

File: dotnet/src/Planners/Planners.Core/Handlebars/IHandlebarsPlanner.cs (hunk 14 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/IHandlebarsPlanner.cs]: You're right, this is really only good for the Stopwatch if we want to leverage that functionality across different planners. I removed this.

File: dotnet/src/Planners/Planners.Core/Handlebars/IHandlebarsPlanner.cs (hunk 15 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/IHandlebarsPlanner.cs]: Removed the interface

**Comment by teresaqhoang** [comment]
> Can we reduce the public API surface to just `HandlebarsPlanner` and `HandlebarsPlan`?
> 
> We should also remove any public interfaces we don't expect customers to implement e.g. `IHandlebarsPlanner`.

@markwallace-microsoft done! I kept HandlebarsPlan and HandlebarsPlanner public but everything else internal

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+using System.Threading.Tasks;
+using Microsoft.SemanticKernel.AI.ChatCompletion;
+using Microsoft.SemanticKernel.Diagnostics;
+
+namespace Microsoft.SemanticKernel.Planners.Handlebars;
+
+/// <summary>
+/// Represents a Handlebars planner.
+/// </summary>
+public class HandlebarsPlanner
```
Comment (alliscode) [inline dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs:18]: Can you make this and the other public classes sealed i.e. `public sealed class HandlebarsPlanner`

**Review by alliscode — APPROVED** [review]

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs (hunk 16 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs]: Done!

**Review by alliscode — APPROVED** [review]

**Comment by teresaqhoang** [comment]
> Found via #3429
> 
> I wonder if a more python like syntax would be easier for LLM to generate as many coding fine tuned LLM has more innate training / understanding for such syntax and would require less handholding via prompt?

We've actually tested with multiple languages, including Python, C#, and TypeScript, and they all run with pretty high accuracy but we're still navigating design decisions on how these plans should be executed! Stay tuned for updates!

File: dotnet/src/Planners/Planners.Core/Handlebars/skPrompt.handlebars (hunk 13 above)
Comment (markwallace-microsoft) [reply dotnet/src/Planners/Planners.Core/Handlebars/skPrompt.handlebars:1]: No, I didn't know that was supported file type. This is very cool.

**Review by alliscode — APPROVED** [review]

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs [hunk 17]
```diff
@@ -0,0 +1,393 @@
...
+            return result.GetValue<object?>();
+        });
+    }
+
+    private static void RegisterSystemHelpers(
+        IHandlebars handlebarsInstance,
+        Dictionary<string, object?> variables
+    )
+    {
+        handlebarsInstance.RegisterHelper("array", (in HelperOptions options, in Context context, in Arguments arguments) =>
```
Comment (markwallace-microsoft) [inline dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs:161]: Question only, are these covered by the standard Handlebars helpers?

From https://github.com/Handlebars-Net/Handlebars.Net/tree/master/source/Handlebars/Helpers

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlannerExtensions.cs [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+        var supplementalNamespace = !additionalNamespace.IsNullOrEmpty() ? $".{additionalNamespace}" : string.Empty;
+        var resourceName = $"{name}{supplementalNamespace}.{fileName}";
+
+        return assembly.GetManifestResourceStream(resourceName)!;
+    }
+
+    /// <summary>
+    /// Start the stopwatch.
+    /// </summary>
+    public static void StartStopwatch(this HandlebarsPlanner planner)
```
Comment (markwallace-microsoft) [inline dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlannerExtensions.cs:50]: These two methods like out of place, why are these needed?

**Review by markwallace-microsoft — APPROVED** [review]

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+        }
+
+        var template = match.Groups[2].Value.Trim(); // match.Success ? match.Groups[2].Value.Trim() : resultContext.Result;
+
+        template = template.Replace("compare.equal", "equal");
+        template = template.Replace("compare.lessThan", "lessThan");
+        template = template.Replace("compare.greaterThan", "greaterThan");
+        template = template.Replace("compare.lessThanOrEqual", "lessThanOrEqual");
+        template = template.Replace("compare.greaterThanOrEqual", "greaterThanOrEqual");
+        template = template.Replace("compare.greaterThanOrEqual", "greaterThanOrEqual");
```
[the remark this answers is not in the document]
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs:93]: To provide some built-in functionality that the model would consistently try to add in the plan. I don't think the latter 4 are as important as equals

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+{
+    /// <summary>
+    /// The key for the available kernel functions.
+    /// </summary>
+    public const string AvailableKernelFunctionsKey = "AVAILABLE_KERNEL_FUNCTIONS";
+
+    /// <summary>
+    /// Gets the stopwatch used for measuring planning time.
+    /// </summary>
+    public Stopwatch Stopwatch { get; } = new();
```
[the remark this answers is not in the document]
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlanner.cs:28]: To track latency of plan creation and execution, this hasn't been implemented yet but will be a fast follow.

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs (hunk 17 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsTemplateEngineExtensions.cs:161]: Ah haven't checked those yet, I'll look through those and remove any duped functionality

File: dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlannerExtensions.cs (hunk 18 above)
Comment (teresaqhoang) [reply dotnet/src/Planners/Planners.Core/Handlebars/HandlebarsPlannerExtensions.cs:50]: Telemetry purposes, will be a fast follow after this merges

**Review by TaoChenOSU — APPROVED** [review]
